frontend/src/components/worksheets/items/TableItem/BundleRow.js

```diff
@@ -0,0 +1,394 @@
+                 className={ classes.buttonsPanel }
+            >
+                <Button
+                    key="upload"
+                    variant="outlined"
+                    size="small"
+                    color="primary"
+                    aria-label="New Upload"
+                    onClick={ () => showNewUpload() }
+                    classes={ { root: classes.buttonRoot } }
+                >
+                    <UploadIcon className={classes.buttonIcon} />
+                    Upload
+                </Button>
+                <Button
+                    key="run"
+                    variant="outlined"
+                    size="small"
+                    color="primary"
+                    aria-label="New Run"
+                    onClick={ () => showNewRun() }
+                    classes={ { root: classes.buttonRoot } }
+                >
+                    <AddIcon className={classes.buttonIcon} />
+                    Run
+                </Button>
+            </div>
+        );
+
+
+    }
+}
+
+class BundleRow extends Component {
+
+    state = {
+        showDetail: false,
+        showNewUpload: 0,
+        showNewRun: 0,
+        showInsertButtons: 0,
+        bundleInfoUpdates: {},
+        showDetail: false,
+        showNewRun: false,
+        showNewUpload: false,
+    }
+
+    receiveBundleInfoUpdates = (update) => {
+        let { bundleInfoUpdates } = this.state;
+        // Use object spread to update.
+        bundleInfoUpdates = {...bundleInfoUpdates, ...update};
+        this.setState({ bundleInfoUpdates });
+    }
+
+    handleClick = () => {
+        this.props.updateRowIndex(this.props.rowIndex);
+        const { showDetail } = this.state;
+        this.setState({
+            showDetail: !showDetail,
+        });
+    }
+
+    showNewUpload = (val) => () => {
+        this.setState({ showNewUpload: val });
+    }
+
+    showNewRun = (val) => () => {
+        this.setState({ showNewRun: val })
+    }
+
+    showButtons = (ev) => {
+        const row = ev.currentTarget;
+        const {
+            top,
+            height,
+        } = row.getBoundingClientRect();
+        const { clientY } = ev;
+        const onTop = (clientY >= top
+                && clientY <= top + 0.25 * height);
+        const onBotttom = (clientY >= top + 0.75 * height
+                && clientY <= top + height);
+        if (onTop) {
+            this.setState({
+                showInsertButtons: -1,
+            });
+        }
+        if (onBotttom) {
+            this.setState({
+                showInsertButtons: 1,
+            });
+        }
+    }
+
+    deleteItem = (ev) => {
+        ev.stopPropagation();
+        const { uuid } = this.props.bundleInfo;
+        $('#command_line')
+            .terminal()
+            .exec(buildTerminalCommand(['rm', uuid]));
+    }
+
+    showMore = (ev) => {
+        ev.stopPropagation();
+    }
+
+    render() {
```

Comment:
There's some perhaps non-intuitive design about how you got the buttons to show up properly.  Document that and what's important to preserve.